### PR TITLE
Android unit test

### PR DESCRIPTION
**Pull request by FabianDevel**
- Fixing old tests in FileControllerTests
- add tests for api calls (In progress)

### CODE REVIEW

File: app/src/main/java/com/infomaniak/drive/data/cache/FileController.kt [hunk 1]
```diff
@@ -712,10 +712,11 @@ object FileController {
     private fun getLocalSortedFolderFiles(
         localFolder: File?,
         order: File.SortType,
-        localChildren: RealmResults<File>? = null
+        localChildren: RealmResults<File>? = null,
+        currRealm: Realm? = null
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/cache/FileController.kt:716]: Rename into `currentRealm`.

File: app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt [hunk 2]
```diff
@@ -257,6 +262,10 @@ object ApiRepository {
         return callApi(ApiRoutes.shareLink(file), PUT, body)
     }
 
+    fun getCategory(driveId: Int): ApiResponse<Array<Category>> {
+        return callApi(ApiRoutes.createCategory(driveId), GET)
+    }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt:267]: `getCategory`, but the Api call is to `createCategory`?

Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt:267]: Oh, it's a `GET` call. We don't have this, do we??

File: app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt [hunk 3]
```diff
@@ -217,6 +218,10 @@ object ApiRepository {
         return callApi(ApiRoutes.unLikeCommentFile(file, commentId), POST)
     }
 
+    fun getShareLink(file: File): ApiResponse<ShareLink> {
+        return callApi(ApiRoutes.shareLink(file), GET)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt:222]: We don't do the GET call for ShareLink, do we?

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+            testFile,
+            body
+        ).also {
+            assertApiResponse(it)
+            Assert.assertEquals("Permission should be public", "public", it.data!!.permission.name.lowercase())
+            Assert.assertFalse("Block downloads should be false", it.data!!.blockDownloads)
+            Assert.assertFalse("Can edit should be false", it.data!!.canEdit)
+            Assert.assertFalse("Show stats should be false", it.data!!.showStats)
+            Assert.assertFalse("Block comments should be false", it.data!!.blockDownloads)
+            Assert.assertFalse("Block information should be false", it.data!!.blockInformation)
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:277]:
```suggestion
            with(it.data!!) {
                Assert.assertEquals("Permission should be public", "public", permission.name.lowercase())
                Assert.assertFalse("Block downloads should be false", blockDownloads)
                Assert.assertFalse("Can edit should be false", canEdit)
                Assert.assertFalse("Show stats should be false", showStats)
                Assert.assertFalse("Block comments should be false", blockDownloads)
                Assert.assertFalse("Block information should be false", blockInformation)
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+        assertApiResponse(response)
+
+        getShareLink(testFile).also {
+            assertApiResponse(it)
+            Assert.assertEquals("Permission should be public", "public", it.data!!.permission.name.lowercase())
+            Assert.assertTrue("block downloads should be true", it.data!!.blockDownloads)
+            Assert.assertTrue("can edit should be true", it.data!!.canEdit)
+            Assert.assertTrue("show stats should be true", it.data!!.showStats)
+            Assert.assertTrue("block comments should be true", it.data!!.blockDownloads)
+            Assert.assertTrue("Block information should be true", it.data!!.blockInformation)
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:304]:
```suggestion
            with(it.data!!) {
                Assert.assertEquals("Permission should be public", "public", permission.name.lowercase())
                Assert.assertTrue("block downloads should be true", blockDownloads)
                Assert.assertTrue("can edit should be true", canEdit)
                Assert.assertTrue("show stats should be true", showStats)
                Assert.assertTrue("block comments should be true", blockDownloads)
                Assert.assertTrue("Block information should be true", blockInformation)
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+
+    @Test
+    fun duplicateTestFile() {
+        val copyName = "test copy"
+        val copyFile = duplicateFile(testFile, copyName, ROOT_ID).also {
+            assertApiResponse(it)
+            Assert.assertEquals("The copy name should be equal to $copyName", copyName, it.data?.name)
+            Assert.assertNotEquals("The id should be different from the original file", testFile.id, it.data?.id)
+            Assert.assertEquals(testFile.driveColor, it.data?.driveColor)
+
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:214]:
```suggestion
            with(it.data!!) {
                Assert.assertEquals("The copy name should be equal to $copyName", copyName, name)
                Assert.assertNotEquals("The id should be different from the original file", testFile.id, id)
                Assert.assertEquals(testFile.driveColor, driveColor)
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+            "email_when_finished" to true,
+            "limit_file_size" to maxSize,
+            "password" to "password"
+        )
+        // Add a dropBox
+        val dropboxId = postDropBox(folder, body).let {
+            assertApiResponse(it)
+            Assert.assertTrue("Email when finished must be true", it.data!!.emailWhenFinished)
+            Assert.assertEquals("Limit file size should be $maxSize", maxSize, it.data!!.limitFileSize)
+            it.data!!.id
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:434]:
```suggestion
            with(it.data!!) {
                Assert.assertTrue("Email when finished must be true", emailWhenFinished)
                Assert.assertEquals("Limit file size should be $maxSize", maxSize, limitFileSize)
                id
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+        }
+
+        getDropBox(folder).also {
+            assertApiResponse(it)
+            Assert.assertEquals("Dropbox name should be '$name'", name, it.data!!.alias)
+            Assert.assertEquals("Dropbox id should be $dropboxId", dropboxId, it.data!!.id)
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:440]:
```suggestion
            with(it.data!!) {
                Assert.assertEquals("Dropbox name should be '$name'", name, alias)
                Assert.assertEquals("Dropbox id should be $dropboxId", dropboxId, id)
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+        updateDropBox(folder, data).also {
+            assertApiResponse(it)
+            Assert.assertTrue(it.data!!)
+        }
+
+        getDropBox(folder).also {
+            assertApiResponse(it)
+            Assert.assertEquals("Dropbox id should be $dropboxId", dropboxId, it.data!!.id)
+            Assert.assertEquals("Email when finished should be false", false, it.data!!.emailWhenFinished)
+            Assert.assertEquals("Limit file size should be ${maxSize * 2}", maxSize * 2, it.data!!.limitFileSize)
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:456]:
```suggestion
            with(it.data!!) {
                Assert.assertEquals("Dropbox id should be $dropboxId", dropboxId, id)
                Assert.assertEquals("Email when finished should be false", false, emailWhenFinished)
                Assert.assertEquals("Limit file size should be ${maxSize * 2}", maxSize * 2, limitFileSize)
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+            fileShareLink.contains("https://drive.infomaniak.com/drive/[0-9]+/file/[0-9]+/share".toRegex())
+        )
+    }
+
+    @Test
+    fun manageCategoryLifecycle() {
+        val categoryResponse = createCategory(userDrive.driveId, "category tests", "#0000FF").also {
+            assertApiResponse(it)
+            Assert.assertEquals("Name of the category should be equals to 'category tests'", "category tests", it.data?.name)
+            Assert.assertEquals("Color of the category should be equals to blue", "#0000FF", it.data?.color)
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:328]:
```suggestion
            with(it.data!!) {
                Assert.assertEquals("Name of the category should be equals to 'category tests'", "category tests", name)
                Assert.assertEquals("Color of the category should be equals to blue", "#0000FF", color)
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+                "Error description should be 'category already exist error'",
+                "Category already exist error",
+                it.error?.description
+            )
+        }
+
+        editCategory(userDrive.driveId, categoryId, "update cat", "#FF0000").also {
+            assertApiResponse(it)
+            Assert.assertEquals("Name of the category should be equals to 'update cat'", "update cat", it.data?.name)
+            Assert.assertEquals("Color of the category should be equals to red", "#FF0000", it.data?.color)
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:345]:
```suggestion
            it.data?.run {
                Assert.assertEquals("Name of the category should be equals to 'update cat'", "update cat", name)
                Assert.assertEquals("Color of the category should be equals to red", "#FF0000", color)
            }
```

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+        createFileForTest().also { file ->
+            val newName = "Trash test"
+            renameFile(file, newName)
+            val modifiedFile = ApiRepository.getLastModifiedFiles(userDrive.driveId).data?.first()
+            Assert.assertNotNull(modifiedFile)
+            deleteTestFile(modifiedFile!!)
+            getTrashFile(modifiedFile, File.SortType.RECENT).also {
+                assertApiResponse(it)
+                Assert.assertEquals("file id should be the same", it.data?.id, file.id)
+                Assert.assertEquals("file name should be updated to 'Trash test'", newName, it.data?.name)
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:476]:
```suggestion
                it.data?.run {
                    Assert.assertEquals("file id should be the same", id, file.id)
                    Assert.assertEquals("file name should be updated to 'Trash test'", newName, name)
                }
```

Comment (JorisBodin) [reply app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:476]: If data is null, the test will not be run, so don't do this

File: app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt [hunk 13]
```diff
@@ -0,0 +1,521 @@
...
+import okhttp3.OkHttpClient
+import org.junit.*
+import org.junit.runner.RunWith
+
+/**
+ * Logging activity testing class
+ */
+@RunWith(AndroidJUnit4::class)
+class ApiRepositoryTest : KDriveTest() {
+    private lateinit var realm: Realm
```
Comment (KevinBoulongne) [inline app/src/androidTest/java/com/infomaniak/drive/ApiRepositoryTest.kt:86]: To check with @sirambd , but this Realm shouldn't exist in this file.

File: app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt (hunk 2 above)
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt:267]: Yes I added it for the tests to be able to assert changes or deletion after calls that don't return the changes

File: app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt (hunk 3 above)
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/drive/data/api/ApiRepository.kt:222]: Same as above, only used for tests

**Review by JorisBodin — APPROVED** [review]

**Review by sirambd — APPROVED** [review]

**Pull request merged**